### PR TITLE
Tool to create SDM files

### PR DESCRIPTION
**Pull request by stefanpsz**
Hi

The Statistical registration needs a binary SDM file in order to work. Given that, I think it would be good to add the tool that creates such files from `mirtk::Matrix` and `mirtk::Vector` files. I also have MATLAB scripts that create such `mirtk::Matrix` and `mirtk::Vector` files from MATLAB matrices and vectors (although some updating might be necessary to work with the new MIRTK), but I do not really know where to put them.

### CODE REVIEW

**Comment by schuhschuh** [comment]
Before even looking into more details into this PR, can you reset your changes create a new commit that does not add all those build tree files such as those in the ```CMakeFiles``` directory and ```CMake/Config``` files ? See last commit 3399c8d. I suggest you use a separate build directory to avoid such mistake.

Would also be nice not to have all those merge commits if anyhow possible. See ```git rebase``` command or ```git pull``` with rebase/fast-forward.

**Comment by ghisvail** [comment]
Please consider rebasing. Just checkout your feature branch and run a `git rebase develop` and fix potential conflicts. Then (force) push to your remote and this PR should be automatically updated.

**Comment by schuhschuh** [comment]
Further, regarding the ```create-sdm``` tool itself, I think this should be part of ```init-dof``` rather. For a statistical FFD, it would take a trained mean vector and covariance matrix as input and write a corresponding statistical FFD ```.dof[.gz]``` file with zero coefficient vector. No need for yet another command.

File: Modules/Transformation/src/BSplineFreeFormTransformationStatistical.cc [hunk 1]
```diff
@@ -42,8 +42,8 @@ ::BSplineFreeFormTransformationStatistical()
 BSplineFreeFormTransformationStatistical
 ::BSplineFreeFormTransformationStatistical(const ImageAttributes    &attr,
                                            CPStatus              ****status,
-                                           const Matrix             &mat,
-                                           const Vector             &vec)
+                                           const mirtk::Matrix      &mat,
```
Comment (schuhschuh) [inline Modules/Transformation/src/BSplineFreeFormTransformationStatistical.cc]: Why did you add the namespace here explicitly? This class is itself in the ```mirtk``` namespace. Please revert this and related changes.

File: Applications/CMakeLists.txt [hunk 2]
```diff
@@ -99,6 +99,7 @@ if (MODULE_Registration)
   endmacro()
 
   add_registration_command(register LibIO OPTIONAL LibLBFGS LibPointSet LibDeformable)
+  add_registration_command(create-sdm LibIO)
```
Comment (schuhschuh) [inline Applications/CMakeLists.txt]: This tool is certainly not a registration command. It does not perform any actual registration. It rather belongs in the transformation command group. Relating to my other comment, it should be an option of the ```init-dof``` command.

File: Applications/create-sdm.cc [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mirtk/Common.h"
+#include "mirtk/Options.h"
+
+#include "mirtk/BaseImage.h"
+#include "mirtk/ImageReader.h"
+#include "mirtk/IOConfig.h"
```
Comment (schuhschuh) [inline Applications/create-sdm.cc]: Why do you include these 3 headers above when this tools does not even read any images?

File: Applications/create-sdm.cc [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+{
+  cout << "\n";
+  cout << "Usage: " << name << "<matrix> <vector> <dofin> <output>\n";
+  cout << "\n";
+  cout << "Description:\n";
+  cout << "  Creates a statistical deformation model (SDM) file to be used in an\n";
+  cout << "  SDM registration procedure.\n";
+  cout << "\n";
+  cout << "Arguments:\n";
+  cout << "  matrix   File containing a mirtkMatrix whose columns are the eigenvectors of the statistical model.\n";
```
Comment (schuhschuh) [inline Applications/create-sdm.cc]: Note that when MIRTK was built WITH_MATLAB, these input files could also be ```.mat``` files. Then you don't need to convert from MATLAB to MIRTK file format when your training is done in MATLAB. Otherwise, when the file name extension of these files is not ```.mat```, assume they are in binary MIRTK Matrix/Vector format.

File: Applications/create-sdm.cc [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  // Read vector
+  if (verbose) cout << "Reading vector file ... "; cout.flush();
+  vector.Read(POSARG(2));
+  if (verbose) cout << "done" << endl;
+  argc--;
+  argv++;
+
+  // Read transformation
+  dofin = Transformation::New(POSARG(3));
+  argc--;
```
Comment (schuhschuh) [inline Applications/create-sdm.cc]: Remove all ```argc--``` and ```argv++```. Did you even test this program ? It wouldn't work like this.

File: Applications/create-sdm.cc [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+  // Create statistical transformation and write SDM
+  BSplineFreeFormTransformationStatistical sffd(ffd->Attributes(), status, matrix, vector);
+
+  if (verbose) cout << "Writing SDM file ... "; cout.flush();
+  sffd.WriteSDM(sdmout_name);
+  if (verbose) cout << "done" << endl;
+
+  // Be nice
+  delete dofin;
```
Comment (schuhschuh) [inline Applications/create-sdm.cc]: Be even nicer and use ```unique_ptr``` instead of a raw C pointer. No need for manual ```delete```.

File: Modules/Transformation/src/BSplineFreeFormTransformationStatistical.cc (hunk 1 above)
Comment (stefanpsz) [inline Modules/Transformation/src/BSplineFreeFormTransformationStatistical.cc]: Regarding the `init-dof` alternative: I believe the `init-dof` tool should output an FFD or DOF. An  SDM file is **not** and FFD nor a DOF. It is the concatenation of attributes, a list of control point status, a matrix and a vector:

```
// -----------------------------------------------------------------------------
void BSplineFreeFormTransformationStatistical::WriteSDM(const char *file)
{
  Cofstream to;
  to.Open(file);

  // Write header with file format version information
  to.WriteAsChar("SDM", 8);
  to.WriteAsChar("v1.0",    5);

  // Write lattice attributes
  to.WriteAsInt(&_attr._x, 1);
  to.WriteAsInt(&_attr._y, 1);
  to.WriteAsInt(&_attr._z, 1);

  to.WriteAsDouble(_attr._xaxis, 3);
  to.WriteAsDouble(_attr._yaxis, 3);
  to.WriteAsDouble(_attr._zaxis, 3);

  to.WriteAsDouble(&_attr._dx, 1);
  to.WriteAsDouble(&_attr._dy, 1);
  to.WriteAsDouble(&_attr._dz, 1);

  to.WriteAsDouble(&_attr._xorigin, 1);
  to.WriteAsDouble(&_attr._yorigin, 1);
  to.WriteAsDouble(&_attr._zorigin, 1);

  to.WriteAsInt(reinterpret_cast<const int *>(_CPStatus[0][0][0]),
                3 * _attr._x * _attr._y * _attr._z);

  // Write statistical deformation model data
  to << _BasisVectors << _MeanVector;
}
```

Given this i think the new tool is needed, because is very specific to SDMs.

With respect to the namespace, I was getting compilation errors:
`no known conversion for argument 4 from ‘mirtk::Vector’ to ‘const Vector& {aka const mirtk::Vector3D<double>&}’`
hence the need to add the namespace in front of Vector. I added it in front of Matrix for stylistic and consistency purposes, but can get rid of it if needed.

File: Applications/CMakeLists.txt (hunk 2 above)
Comment (stefanpsz) [inline Applications/CMakeLists.txt]: I can centainly put it somewhere else, but as I mention above, it has nothing to do with FFD or DOFs

File: Applications/create-sdm.cc (hunk 3 above)
Comment (stefanpsz) [inline Applications/create-sdm.cc]: Mistake

File: Modules/Transformation/src/BSplineFreeFormTransformationStatistical.cc (hunk 1 above)
Comment (schuhschuh) [inline Modules/Transformation/src/BSplineFreeFormTransformationStatistical.cc]: > An SDM file is not and FFD nor a DOF.

That's what I think is maybe not the best way to do it. A SDM could be considered a Statistical FFD. One without coefficient vector because this vector is all zero (i.e., a sparse representation of this vector which does not take up any space). The input to a SDM based registration would be this transformation file as ```-dofin``` and the output ```-dofout``` is one with a non-zero coefficient vector.

Comment (schuhschuh) [inline Modules/Transformation/src/BSplineFreeFormTransformationStatistical.cc]: > no known conversion for argument 4 from ‘mirtk::Vector’ to ‘const Vector& {aka const mirtk::Vector3D<double>&}’

A better solution would be to use a more specific name for the ```Vector``` and ```Matrix``` types:
```c++
typedef ... MeanVector;
typedef ... CovarianceMatrix;
```

File: Applications/CMakeLists.txt (hunk 2 above)
Comment (schuhschuh) [inline Applications/CMakeLists.txt]: See my other comment. I don't think so. a SDM has a lot to do with your statistical FFD. It **is** a model for such FFD, or, one particular instance of it.

**Comment by schuhschuh** [comment]
Please note that Applications directory changes of PR #168 which require rebasing your branch and resolving the conflicts.

**Comment by stefanpsz** [comment]
Since the requested changes to this PR are way beyond its initial intended scope, I opted to close it. Will create another branch in my repository and create a new PR with the changes in the future.

**Comment by schuhschuh** [comment]
Thanks for clarifying, though I thought already that's why. When you get back to this, it is probably a good idea to discuss the specifics how to incorporate the statistical registration tools best in the current set of tools in the thread of the original feature request issue #66.
